extendedcopy.go

```diff
@@ -170,3 +173,48 @@ func findRoots(ctx context.Context, storage content.GraphStorage, node ocispec.D
 	}
 	return roots, nil
 }
+
+// FilterArtifactType will configure opts.FindPredecessors to filter the predecessors
+// whose artifact type matches a given regex pattern. The regex syntax supported
+// is RE2. Reference: https://github.com/google/re2/wiki/Syntax.
+func (opts *ExtendedCopyGraphOptions) FilterArtifactType(pattern string) {
+	fp := opts.FindPredecessors
+	opts.FindPredecessors = func(ctx context.Context, src content.GraphStorage,
+		desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		regex, err := regexp.Compile(pattern)
+		if err != nil {
+			return nil, err
+		}
+		var predecessors []ocispec.Descriptor
+		if fp == nil {
+			predecessors, err = src.Predecessors(ctx, desc)
+		} else {
+			predecessors, err = fp(ctx, src, desc)
+		}
+		if err != nil {
+			return nil, err
+		}
+		var filtered []ocispec.Descriptor
+		for _, p := range predecessors {
+			if p.MediaType == artifactspec.MediaTypeArtifactManifest {
+				rc, err := src.Fetch(ctx, p)
+				if err != nil {
+					return nil, err
+				}
+				defer rc.Close()
```

Comment:
`defer` in a for loop is dangerous. The deferred function will not be executed at the end of each iteration of the for loop but when the function is returned.

Reply:
Wrapped defer with an anonymous function.